crypto/bio/bio_lib.c

```diff
@@ -45,22 +73,33 @@ int BIO_set(BIO *bio, const BIO_METHOD *method)
     bio->references = 1;
     bio->num_read = 0L;
     bio->num_write = 0L;
+}
+#endif
+
+static int bio_init(BIO *bio, const BIO_METHOD *method)
+{
+    bio->method = method;
+    bio->shutdown = 1;
+    bio->references = 1;
+
     if (!CRYPTO_new_ex_data(CRYPTO_EX_INDEX_BIO, bio, &bio->ex_data))
         return 0;
 
     bio->lock = CRYPTO_THREAD_lock_new();
     if (bio->lock == NULL) {
-        BIOerr(BIO_F_BIO_SET, ERR_R_MALLOC_FAILURE);
+        BIOerr(BIO_F_BIO_INIT, ERR_R_MALLOC_FAILURE);
         CRYPTO_free_ex_data(CRYPTO_EX_INDEX_BIO, bio, &bio->ex_data);
+        bio->ex_data.sk = NULL;
         return 0;
     }
 
-    if (method->create != NULL) {
-        if (!method->create(bio)) {
-            CRYPTO_free_ex_data(CRYPTO_EX_INDEX_BIO, bio, &bio->ex_data);
-            CRYPTO_THREAD_lock_free(bio->lock);
-            return 0;
-        }
+    if (method->create && !method->create(bio)) {
```

Comment:
No?? I meant:
`method->create != NULL && !method->create(bio)`

The return value from the `create` function call really is a boolean, so its fine to treat it as such.